server/src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java

```diff
@@ -134,24 +153,37 @@ void cleanupAllocatedDangledIndices(Metadata metadata) {
      * to the currently tracked dangling indices.
      */
     void findNewAndAddDanglingIndices(final Metadata metadata) {
-        danglingIndices.putAll(findNewDanglingIndices(metadata));
+        final IndexGraveyard graveyard = metadata.indexGraveyard();
+
+        // If a tombstone is created for a dangling index, we need to make sure that the
+        // index is no longer considered dangling.
+        for (Index key : danglingIndices.keySet()) {
+            if (graveyard.containsIndex(key)) {
+                danglingIndices.remove(key);
+            }
+        }
+
+        danglingIndices.putAll(findNewDanglingIndices(danglingIndices, metadata));
     }
 
     /**
      * Finds new dangling indices by iterating over the indices and trying to find indices
      * that have state on disk, but are not part of the provided meta data, or not detected
      * as dangled already.
      */
-    Map<Index, IndexMetadata> findNewDanglingIndices(final Metadata metadata) {
+    public Map<Index, IndexMetadata> findNewDanglingIndices(Map<Index, IndexMetadata> existingDanglingIndices, final Metadata metadata) {
```

Comment:
Should probably have a unit test that passes something other than `emptyMap()` to this new parameter.